[자동차 경주 미션] 조태현 미션 제출합니다.

whxogus215 opened the pull request:
# 패키지 구조
- `controller` : `main` 메서드에서 호출하는 게임 실행 클래스
- `domain` : 자동차 경주 게임에 필요한 비즈니스 로직이 담긴 클래스들
- `exception` : 자동차 경주에서 발생할 수 있는 커스텀 예외 및 에러 메시지
- `utils` : 도메인과 관련은 없지만 게임 실행에 있어서 부가적인 기능들을 제공하는 클래스들
- `validator` : 입력 값을 검증하는 클래스 (utils와 비슷한 느낌이 있으나, 검증이라는 특징을 드러내고자 패키지를 분리해보았습니다.)
- `view` : 게임 실행에 필요한 입,출력 값들을 처리하는 클래스들

# MVC 구조
- 4단계 리팩토링 힌트를 참고하면서 도메인과 그 외의 것들로 분리하고자 하였습니다.
  - 이들을 분리함으로써 테스트해야 하는 부분이 어떤 것들인지 명확하게 나눠질 수 있었습니다.
 
# 같이 의논해보고 싶은 부분
- **랜덤 값을 생성하는 로직에 대한 테스트**를 과연 해야하는 것이 맞을까 하는 의문이 계속 들었습니다.
  - 페어 프로그래밍을 하면서도 자바가 제공하는 API를 테스트하는 것이 합리적인가에 대해 의논을 해보았으나 마땅한 결과는 얻지 못했습니다.
  - 리뷰어님의 생각은 어떠신지 궁금합니다😄

mete0rfish committed:
자동차 경주 1단계 완료

mete0rfish committed:
자동차 경주 2단계 완료

whxogus215 committed:
docs: 리드미 초안 작성

whxogus215 committed:
feat: 자동차 이름에 대한 검증 메서드 구현 및 테스트

- 이름이 6자 이상일 때, 예외 발생
- 이름이 1자 미만일 때, 예외 발생
- @ParameterizedTest, @ValueSource로 여러 파라미터에 대한 테스트 작성

whxogus215 committed:
feat: 사용자의 입력 값을 받는 역할 구현

whxogus215 committed:
feat: 자동차 이름을 리스트로 반환하는 파서 구현

whxogus215 committed:
feat: 경주 관련 메서드 수정

- 우승자 선정(selectWinners()) : 우승자의 이름만 담겨있는 리스트로 반환
- 경주 시작(raceStart()) : 게임 횟수만큼 이동거리를 갱신하고 그 결과 값을 출력

whxogus215 committed:
feat: View 객체 구현

- GamePrinter: 랩별 결과값, 우승자 출력
- GameReader: 경기에 필요한 이름, 횟수 입력

whxogus215 committed:
feat: 실행 프로그램 구현

whxogus215 committed:
docs: 리드미 업데이트

whxogus215 committed:
feat: 입력 값에 대한 검증 객체 구현

- 자동차 이름이 중복되는지 확인
- 게임 횟수가 1 미만인지 확인

whxogus215 committed:
style: 위아래 공백 수정

whxogus215 committed:
3단계 미션 완료

whxogus215 committed:
feat: 예외 패키지 및 커스텀 예외, 에러 메시지 Enum 추가

whxogus215 committed:
refactor: 자동차 생성 시 예외 발생 테스트 리팩토링

whxogus215 committed:
refactor: MVC를 모방한 패키지 분리

whxogus215 committed:
refactor: 메서드 접근제어자 변경

whxogus215 committed:
refactor: 컨트롤러 메서드 구현 및 main 메서드 간소화

whxogus215 committed:
refactor: 랜덤 값을 생성하는 로직을 domain과 느슨하게 결합하도록 리팩토링

- 함수형 인터페이스를 GameRace의 로직에 추가함으로써 추상화
- GameController에서 입력 값의 형태를 자유롭게 바꿀 수 있도록 변경
- 그에 맞게 원하는 값을 넣어서 테스트하는 테스트 코드를 작성

whxogus215 committed:
test: 우승자 리스트 반환 테스트

whxogus215 committed:
refactor: CarRace[domain]가 GamePrinter[view]에 의존하지 않도록 변경

whxogus215 committed:
refactor: CarException 커스텀 예외를 발생하도록 리팩토링

whxogus215 committed:
refactor: CarTest 페어프로그래밍 때 작성한 코드 리팩토링

- AssertJ의 assertThat, isEqualTo 사용

whxogus215 committed:
코드 검토 완료 및 4단계 미션 완료

TaeyeonRoyce commented on README.md:
````diff
@@ -0,0 +1,89 @@
+neo, brown가 최종 우승했습니다.
+```
+
+## 제출 전 체크
+- [x] 클래스 별 코드 위아래 공백 확인
+- [x] 기존 프로그래밍 요구사항 준수하는지 확인
````
코드를 체크하는 태현님만의 체크리스트가 있군요👍
배워갑니다~!

TaeyeonRoyce commented on src/main/java/controller/CarController.java:
```diff
@@ -0,0 +1,55 @@
+    }
+
+    // 자동차 이름 입력받기
+    private List<Car> inputCarNames() {
```
시그니처만 봤을때 혼동을 줄 수도 있지 않을까 싶습니다.
`CarNames`가 아닌 `List<Car>`를 반환하고 있는데요.
현재 메서드 네임으로는 `CarNames`가 반환 되는게 자연스러워 보이지 않으신가요?

입력을 읽는 것과 자동차 생성이라는 작업을 나누어 볼 수도 있겠네요.

whxogus215 replied:
컨트롤러에 실제 도메인 객체를 생성하는 로직이 포함되어 있다보니 역할이 분명하게 나눠지지 않은게 원인인 것 같습니다. `inputCarNames`인만큼 반환타입이 `List<String>`이 좀 더 알맞은 것 같습니다. 다음부터는 메서드 시그니처와 실제 구현부가 얼만큼 일치하는지도 잘 파악해야할 것 같아요!

TaeyeonRoyce commented on src/main/java/controller/CarController.java:
```diff
@@ -0,0 +1,55 @@
+    // 자동차 이름 입력받기
+    private List<Car> inputCarNames() {
+        String carNames = GameReader.getNames();
+        List<String> parsedNames = Parser.parseStringArray(carNames);
+        List<Car> cars = new ArrayList<>();
+        for (String name : parsedNames) {
+            cars.add(new Car(name));
+        }
+        return cars;
```
저는 반복성 로직에 대해 최대한 `stream`을 활용하려고 해요.
가독성도 높고, 불필요한 초기화도 필요하지 않아서요!
태현님은 어떤걸 선호하시는 편인가요?

아래처럼 활용할 수 있겠네요!
```suggestion
        List<String> parsedNames = Parser.parseStringArray(carNames);
        return parsedNames.stream()
                .map(Car::new)
                .toList();
```

whxogus215 replied:
분명 다른 코드들에서는 스트림을 활용했다고 생각했는데 빼먹은 부분이 있었네요! 깔끔하고 좋은 코드 감사합니다

TaeyeonRoyce commented on src/main/java/domain/Car.java:
```diff
@@ -0,0 +1,45 @@
+
+public class Car {
+
+    private final int MIN_LENGTH = 1;
+    private final int MAX_LENGTH = 5;
```
상수로 정의 했다면, static으로 선언하시죠!
```suggestion
    private static final int MIN_LENGTH = 1;
    private static final int MAX_LENGTH = 5;
```

static으로 선언하지 않은것과는 어떤차이가 있을까요?

whxogus215 replied:
static으로 선언하는 이유는 클래스 간 공통으로 사용하는 변수를 만들 수 있기 때문입니다. 자동차 이름의 최소/최대 길이는 인스턴스마다 별도의 값을 가져선 안되기 때문에 상수로 정의한 부분은 static이 올바른 것 같습니다.

TaeyeonRoyce commented on src/main/java/exception/CarException.java:
```diff
@@ -0,0 +1,13 @@
+package exception;
+
+public class CarException extends IllegalArgumentException{
```
CarExcpetion을 구현하고, 메세지로 케이스를 분류하셨군요.👍

whxogus215 replied:
예외 학습테스트를 진행하면서 커스텀 예외의 중요성을 알게 되어 적용해보았습니다!

TaeyeonRoyce commented on src/main/java/utils/Parser.java:
```diff
@@ -0,0 +1,26 @@
+import java.util.List;
+import validator.InputValidator;
+
+public class Parser {
```
`Parser`는 다양하게 사용될 수 있는 범용어에요.
해당 클래스의 역할에 맞게 구체적으로 표현하면 어떨까요?
`CarNameParser`가 떠오르네요!

whxogus215 replied:
그러게요. `Parser`라는 이름을 붙이면서도 뭔가 추상적이다라는 느낌이 들었는데, 좀 더 구체적인 수식어가 필요할 것 같네요!

TaeyeonRoyce commented on src/main/java/utils/Parser.java:
```diff
@@ -0,0 +1,26 @@
+
+    private static List<String> parseStringArrayWithComma(String names) {
+        if (names == null || names.isEmpty()) {
+            return new ArrayList<>(); // 빈 배열 반환
```
```suggestion
            return Collections.emptyList(); // 빈 배열 반환
```
컬렉션의 정적메서드로도 사용할 수 있어요.
`new ArrayList<>();` 와는 어떤 차이가 있을까요?

whxogus215 replied:
`Collections.emptyList()`는 `ArrayList`가 아닌 `EmptyList`라는 불변(immutable) 객체를 반환하고 있네요! 또한 메서드 이름만보더라도 빈 리스트를 반환한다는 의도가 잘 전달되는 것 같아요

TaeyeonRoyce commented on src/main/java/validator/InputValidator.java:
```diff
@@ -0,0 +1,26 @@
+    private static final int MIN_ROUND_COUNT = 1;
+
+    public static void validateDuplicate(List<String> input) {
+        Set<String> convertedInput = new HashSet<>(input);
+        if (input.size() > convertedInput.size()) {
+            throw new CarException(ErrorMessage.NAME_NOT_DUPLICATED);
+        }
```
이름 중복 검사까지👍

TaeyeonRoyce commented on src/main/java/view/GamePrinter.java:
```diff
@@ -0,0 +1,27 @@
+    }
+
+    public static void printRaceWinners(List<String> winnerNames) {
+        StringJoiner joiner = new StringJoiner(", ");
```
`StringJoiner`를 활용하여 출력 로직이 더 간편해졌네요👍

TaeyeonRoyce commented on src/test/java/RandomValueGeneratorTest.java:
```diff
@@ -0,0 +1,20 @@
+    @DisplayName("0부터 9까지 랜덤 숫자를 생성하는지 확인한다.")
+    void generateZeroToNineTest() {
+        int value = RandomValueGenerator.generate();
+        assertFalse(value < 0 || value >= 10);
```
다른 테스트 코드들 처럼 검증부에서 `AssertJ`를 활용하면 어떨까요?
`assertThat().isBetween();` 을 활용해볼 수 있겠네요!

whxogus215 replied:
좋은 메서드 감사합니다! 훨씬 더 의미가 잘 전달되네요

TaeyeonRoyce commented on src/test/java/domain/CarRaceTest.java:
```diff
@@ -0,0 +1,82 @@
+                                new Car("페라리"),
+                                new Car("람보르기니"),
+                                new Car("애스턴마틴"),
+                                new Car("벤틀리")
```
좋은 차 타시네욥

TaeyeonRoyce commented on src/test/java/domain/CarRaceTest.java:
```diff
@@ -0,0 +1,82 @@
+
+        int gameCount = 5;
+        for (int i = 0; i < gameCount; i++) {
+            carRace.raceOneLap(() -> 7);
```
람다 활용👍

TaeyeonRoyce commented on src/test/java/domain/CarRaceTest.java:
```diff
@@ -0,0 +1,82 @@
+    void selectWinnersTest(List<Car> cars) {
+        CarRace carRace = new CarRace(cars);
+
+        int gameCount = 5;
+        for (int i = 0; i < gameCount; i++) {
+            carRace.raceOneLap(() -> 7);
+        }
+        List<String> winners = carRace.selectWinners();
+
+        assertThat(winners)
+                .containsExactlyInAnyOrderElementsOf(cars.stream()
+                        .map(Car::getName)
+                        .collect(Collectors.toList()));
```
작성한 로직을 제대로 검증하는지 모르겠어요.
현재 방식은 모든 자동차가 우승하고 있는데요.

랜덤성 때문에 어려워 보이는데 어떻게 해결할 수 있을까요?
간단하게는 람다식의 int를 다르게 주는 방식이 생각납니다..!

whxogus215 replied:
의도한 값을 전달할 수 있도록 `ValueGenerator`라는 함수형 인터페이스를 도입했었는데, 역시 테스트하기가 쉽지 않네요ㅎㅎ;

TaeyeonRoyce commented on src/test/java/domain/CarTest.java:
```diff
@@ -0,0 +1,64 @@
+    void moveTest() {
+        Car car = new Car("포르쉐");
+        int beforeDistance = car.getDistance();
+        for (int i = 4; i < 10; i++) {
+            car.moveOrStop(i);
+            assertThat(car.getDistance()).isEqualTo(beforeDistance + 1);
+            beforeDistance = car.getDistance();
+        }
```
하나의 검증 메서드에서 케이스에 따른 여러번 검증을 거치는 것 보단,
`@ValueSource(ints = {})`를 활용하는건 어떨까요?

테스트코드는 단순할수록 유용하다고 생각해요.
프로덕션 로직의 동작을 테스트로 표현 할 수도 있는데, 해당 테스트 코드도 많은 해석이 필요하다면 불편하겠죠.

whxogus215 replied:
`@ParameterizedTest`와 `@ValueSource`를 적극적으로 활용해보도록 하겠습니다!

TaeyeonRoyce commented on src/test/java/domain/CarTest.java:
```diff
@@ -0,0 +1,64 @@
+        int beforeDistance = car.getDistance();
+        car.moveOrStop(value);
+
+        if (value >= 4 && value <= 9) {
+            assertThat(car.getDistance()).isEqualTo(beforeDistance + 1);
+        }
+        if (value < 4 || value > 9) {
+            assertThat(car.getDistance()).isEqualTo(beforeDistance);
+        }
```
랜덤성 때문에 테스트의 검증이 분리되었네요.
테스트는 언제 어디서 수행해도 같은 결과가 나와야 합니다.
실행 할때마다 통과하거나 실패한다면 의미가 없죠.

현재 코드는 항상 성공하는 코드이지만, 하위 조건문(`value < 4 || value > 9`)이 항상 실행된다고 보장할 수 있을까요?
추후 수정하다보면 버그가 들어갈 수 있는 영역이 될거라 생각이 드네요.

어떻게 하면 랜덤성을 효과적으로 테스트할 수 있을까요?
랜덤성도 테스트의 대상일까요?
해당 주제로 고민해보시면 좋겠습니다!

whxogus215 replied:
사용자가 의도한 값을 전달할 수 있도록, 값을 주입받는 부분을 추상화함으로써 원하는 값을 전달할 수 있도록 하는 방법이 제일 좋을 것 같다고 생각했습니다. 제가 작성했던 `CarRace`의 `raceOneLap` 메서드처럼 외부에 의존해야하는 기능을 추상화한다면 의도한 랜덤 값을 줄 수 있을 것 같습니다!

TaeyeonRoyce reviewed (commented):
안녕하세요, 태현님!
고민이 엿보이는 코드인 구조도 잘 나누어져있고, 전반적으로 깔끔하네요!
잘 정리된 기능명세서도 인상깊었습니다.
고생 많으셨어요👍👍

몇가지 코멘트 남겨두었는데 확인하시고 답변주세요!

---
코멘트 답변

남겨주신 의논하고 싶은 내용에 대한 생각을 말씀드릴게요.
자바가 제공하는 API를 테스트 해야한다면 할 대상들이 너무 많아요..ㅎ
저는 자바 API만큼 신뢰도 높은 API나, 필연적인 API에 대해선 직접적인 호출을 통한 테스트는 하지 않는 편입니다.
테스트의 목적을 기준으로 판단하면 좋겠네요.
우리가 테스트하고 싶은 관심사 혹은 대상은 저희가 쌓은 로직(논리)이니까요.

하지만, 외부 라이브러리의 경우 방어 코드(테스트)는 반드시 사용해요.
동작에 대한 신뢰도 뿐만 아니라, 버전 변경에 따른 영향도도 파악 할 수 있어서요.
태현님이 체크리스트를 만든 것처럼, 본인 만의 기준을 정립해보면 어떨까요?
저는 아래와 같은 기준들이 생각나네요.
1. 신뢰 할 수 있는 프로젝트인가?(사용자, star, download 수 등등의 지표)
2. 내 코드가 해당 라이브러리에 얼마나 의존적인가?
3. 대체할 수 있는 라이브러리가 있는가?
4. 사용하는데 학습 곡선이 높지 않은가?
등등..

답변이 되었으면 좋겠습니다.
또 궁금하시거나 의논하고 싶으신 부분 있으시면 남겨주세요.
고생하셨습니다👍

whxogus215 commented:
@TaeyeonRoyce 당장 미션을 진행하면서도 자바의 여러 API를 사용하고 있는데 이들에 대한 신뢰를 하지 못하는 모순점이 있었네요 ㅎㅎ 제시해주신 여러 기준이 구체적이어서 와닿았던 것 같아요! 저도 저만의 기준을 세워보도록 하겠습니다 👍

whxogus215 committed:
feat: 자동차 생성하는 역할을 지닌 CarFactory 객체 추가

- 변경 전 : carNames라는 변수이름을 가졌음에도 객체 타입이 List<Car>이어서 직관적이지 않음.
- 변경 후 : carNames는 List<String>, cars는 List<Car>로 변수이름과 객체 타입이 논리적으로 일치하도록 함.
- 자동차 생성을 하는 역할을 추가함으로써 자동차 생성의 조건이 변동되더라도 컨트롤러는 영향을 받지 않게 됨.

whxogus215 committed:
refactor: 상수 값 static으로 변경

- 클래스 변수로 선언함으로써 각 인스턴스가 공통으로 갖게되는 데이터로 만듦

whxogus215 committed:
refactor: Parser -> CarNameParser 클래스명 변경

- Parser라는 이름 갖고는 정확히 어떤 데이터를 변환하는 건지 알기가 어려움. 따라서 앞에 수식어를 붙여서 의미를 구체적으로 표현함

whxogus215 committed:
refactor: 빈 배열 반환하는 메서드로 변경

- Collections의 emptyList()를 사용하면, 불변 리스트를 반환함과 동시에 메서드 이름만으로도 무얼 반환하는지 알 수 있음.

whxogus215 committed:
refactor: 테스트하는 값이 특정 범위에 있는지 확인하는 AssertJ 메서드로 변경

whxogus215 committed:
refactor: 랜덤 값을 반환하는 람다식 간소화(기존에 생성한 Generator 사용)

whxogus215 committed:
feat: 자동차의 필드 값으로 Generator 인터페이스 추가

- 자동차 생성 시 값을 생성하는 방식을 지정하도록 변경 -> 랜덤 값을 테스트하기 용이하도록 함.
  - 테스트 대상이 되는 각 자동차에 대해 일일이 지정할 수 있기 때문
- 생성자 및 메서드의 매개변수에 final 붙여서 코드의 안정성 증가

whxogus215 committed:
fix: 리팩토링 후 컴파일 에러 수정

whxogus215 committed:
refactor: Car 도메인 테스트 코드 리팩토링

whxogus215 committed:
refactor: CarRace 도메인 테스트 코드 리팩토링 및 final 추가
